smsjwplatform/acl.py

```diff
@@ -0,0 +1,182 @@
+Module providing lookup API-related functionality.
+
+"""
+import logging
+from urllib.parse import urljoin
+from django.conf import settings
+from django.core.cache import cache
+
+from .oauth2client import AuthenticatedSession
+
+LOG = logging.getLogger(__name__)
+
+
+#: An authenticated session which can access the lookup API
+LOOKUP_SESSION = AuthenticatedSession(scopes=settings.SMS_OAUTH2_LOOKUP_SCOPES)
+
+
+def get_person_for_user(user):
+    """
+    Return the resource from Lookup associated with the specified user. A requests package
+    :py:class:`HTTPError` is raised if the request fails.
+
+    The result of this function call is cached based on the username so it is safe to call this
+    multiple times.
+
+    If user is the anonymous user (user.is_anonymous is True), :py:class:`~.UserIsAnonymousError`
+    is raised.
+
+    """
+    # check that the user is not anonymous
+    if user.is_anonymous:
+        raise RuntimeError('User is anonymous')
+
+    # return a cached response if we have it
+    cached_resource = cache.get(f"{user.username}:lookup")
+    if cached_resource is not None:
+        return cached_resource
+
+    # Ask lookup about this person
+    lookup_response = LOOKUP_SESSION.request(
+        method='GET', url=urljoin(
+            settings.LOOKUP_ROOT,
+            f'people/{settings.LOOKUP_PEOPLE_ID_SCHEME}/{user.username}?fetch=all_insts,all_groups'
+        )
+    )
+
+    # Raise if there was an error
+    lookup_response.raise_for_status()
+
+    # save cached value
+    cache.set(f"{user.username}:lookup", lookup_response.json(),
+              settings.LOOKUP_PEOPLE_CACHE_LIFETIME)
+
+    # recurse, which should now retrieve the value from the cache
+    return get_person_for_user(user)
+
+
+class AceWorld:
+    """This class encapsulates an ACE of the form WORLD"""
+
+    @staticmethod
+    def parse(ace):
+        """Parses an ACE and constructs/return an AceWorld if it matches or None if it doesn't"""
+        return AceWorld() if ace == 'WORLD' else None
+
+    @staticmethod
+    def has_permission(user):
+        """Anyone can see the media"""
+        LOG.debug(user)
+        return True
+
+
+class AceCam:
+    """This class encapsulates an ACE of the form CAM"""
+
+    @staticmethod
+    def parse(ace):
+        """Parses an ACE and constructs/return an AceCam if it matches or None if it doesn't"""
+        return AceCam() if ace == 'CAM' else None
+
+    @staticmethod
+    def has_permission(user):
+        """Any authenticated user can see the media"""
+        return not user.is_anonymous
+
+
+class AceInst:
+    """This class encapsulates an ACE of the form INST_{instid}"""
+
+    prefix = 'INST_'
+
+    def __init__(self, instid):
+        self.instid = instid
+
+    @staticmethod
+    def parse(ace):
+        """Parses an ACE and constructs/return an AceInst if it matches or None if it doesn't"""
+        return AceInst(ace[len(AceInst.prefix):]) if ace.startswith(AceInst.prefix) else None
```

Comment:
Very minor niggle: I think ``ace.partition(AceInst.prefix)[2]`` is slightly nicer. One could, perhaps even use partition for the entire thing:
```python
@staticmethod
def parse(ace):
    """Parses an ACE and constructs/return an AceInst if it matches or None if it doesn't"""
    head, prefix, tail = ace.partition(AceInst.prefix)
    return AceInst(tail) if head == '' and prefix == AceInst.prefix else None
```

This is entirely subjective, of course, so no harm, no foul if you think ``ace[len(AceInst.prefix):]`` is nicer.